### PR TITLE
add implementation of missing tiles generator

### PR DESCRIPTION
**Pull request by peitili**
Add bounding box rebuild ability

### Background
After each global build, we might want to rebuild the area in a certain bounding box, so adding the ability to achieve that.

### Test
Triggered the build using a few bounding boxes and verified the tiles manually.

### CODE REVIEW

File: batch-setup/make_meta_tiles.py [hunk 1]
```diff
@@ -457,15 +484,25 @@ def render(self, num_retries, lense):
     assert args.metatile_size > 0
     assert args.metatile_size < 100
     metatile_max_zoom = 16 - metatile_zoom_from_size(args.metatile_size)
+
+    generator = None
+    if args.use_tiles_coords_generator:
+        generator = BoundingBoxTilesCoordinateGenerator(west=-4.1494623,
```
Comment (nvkelso) [inline batch-setup/make_meta_tiles.py:490]: Please set the default bbox to be over Seattle, SF, or NYC.

File: batch-setup/make_meta_tiles.py [hunk 2]
```diff
@@ -145,34 +175,28 @@ def missing_tiles_split(self, split_zoom, zoom_max, big_jobs):
...
         try:
-            missing_meta_file = os.path.join(tmpdir, 'missing_meta.txt')
             missing_low_file = os.path.join(tmpdir, 'missing.low.txt')
             missing_high_file = os.path.join(tmpdir, 'missing.high.txt')
 
-            self.read_metas_to_file(missing_meta_file, compress=True)
-
             print("[make_meta_tiles] Splitting into high and low zoom lists")
 
             # contains zooms 0 until group zoom. the jobs between the group
```
Comment (peitili) [inline batch-setup/make_meta_tiles.py:183]: this comment is really confusing? the code below is saying `zoom_max` the comment here refers to `group zoom`, what is the relationship of these two terms?

File: batch-setup/make_meta_tiles.py (hunk 1 above)
Comment (peitili) [reply batch-setup/make_meta_tiles.py]: > Please set the default bbox to a wide region over Seattle, SF, or NYC.

removed the hardcoded bbox

File: utils/tiles.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+from ModestMaps.Core import Coordinate
+from mercantile import tiles
+from mercantile import tile
+#from typing import Iterator
```
Comment (travisgrigsby) [inline utils/tiles.py:4]: remove commented imports

**Review by travisgrigsby — CHANGES_REQUESTED** [review]
This is looking great!  A bunch of minor comments.  Just one thing I'm concerned about:
Am I missing the part where you actually verify that the tiles were created?

File: utils/tiles.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#from typing import List
+from utils.constants import MAX_TILE_ZOOM
+from utils.constants import MIN_TILE_ZOOM
+
+
+class TilesCoordinateGenerator(object):
```
[the remark this answers is not in the document]
Comment (peitili) [reply utils/tiles.py:10]: > are you using this base class anywhere?

no it is for future extending considering the following feature


`in addition to a bounding box, we’ll want to eventually support a tile list too - that’s the way diff builds will work.  So the bounding box could drive the same code path if you turn it into a list of tiles first, which is probably what you’re doing.`

File: utils/tiles.py (hunk 3 above)
Comment (peitili) [reply utils/tiles.py:4]: > remove commented imports

sure. But some background, but why it is here is that in my local python3 environment, I used it for type checking, it is way better experience in terms of development.

However in our runtime environment, python2, it is not supported so I commented them out, I can completely remove them, that's fine.

File: batch-setup/make_meta_tiles.py [hunk 5]
```diff
@@ -130,8 +138,36 @@ def read_metas_to_file(self, filename, present=False, compress=False):
...
+        """ generate the missing tiles coordinates for low/high-zoom
+            splitting
+            :param try_generator: if set, it will to use the preset tiles_coords_generator if that's available
+        """
+        tmpdir = tempfile.mkdtemp()
+        try:
+            if try_generator and bool(self.tiles_coords_generator):
+                print("[make_meta_tiles] generate missing tiles coords "
+                      "using customized generator")
+                # todo check how yield can return
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_meta_tiles.py:152]: > remove todo

removed in the next commit.

File: batch-setup/make_meta_tiles.py [hunk 6]
```diff
@@ -130,8 +138,36 @@ def read_metas_to_file(self, filename, present=False, compress=False):
...
+                # for coord in self.tiles_coords_generator.generate_tiles_coordinates([]):
+                #     yield coord
+                return [coord for coord in self.tiles_coords_generator.generate_tiles_coordinates([])]
+            else:
+                self.run_batch_job()
+                missing_meta_file = os.path.join(tmpdir, 'missing_meta.txt')
+                self.read_metas_to_file(missing_meta_file, compress=True)
+                with gzip.open(missing_meta_file, 'r') as fh:
+                    # for line in fh:
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_meta_tiles.py:161]: > remove comment

removed in the next commit.

File: batch-setup/make_meta_tiles.py [hunk 7]
```diff
@@ -130,8 +138,36 @@ def read_metas_to_file(self, filename, present=False, compress=False):
                '-max-zoom', str(self.max_zoom),
                stdout=filename)
 
+    def generate_missing_tiles_coords(self, try_generator=False):
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_meta_tiles.py:141]: > I think you can do a global find and replace on `*tiles_coords*` -> `*tile_coords*`

fixed in the next commit?

File: batch-setup/make_meta_tiles.py [hunk 8]
```diff
@@ -341,23 +371,25 @@ def missing_file(self, missing):
...
-            self.split_zoom, self.zoom_max, self.big_jobs)
+            self.split_zoom, self.zoom_max, self.big_jobs, try_generator)
 
     def render(self, num_retries, lense):
         mem_max = 32 * 1024  # 32 GiB
 
         for retry_number in range(0, num_retries):
             mem_multiplier = 1.5 ** retry_number
-            with self._missing() as missing:
+            try_generator = (retry_number == 0)
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_meta_tiles.py:391]: Note that the following is not true. For force override to happen, two conditions have to be met, `try_generator==True` and  `tile_coords_generator is not None` ; so the original behavior is still there if you don't specify a tile_coords_generator. 

> isn't going to happen anymore - we are forcing an overwrite

File: batch-setup/make_meta_tiles.py [hunk 9]
```diff
@@ -457,15 +494,36 @@ def render(self, num_retries, lense):
     assert args.metatile_size > 0
     assert args.metatile_size < 100
     metatile_max_zoom = 16 - metatile_zoom_from_size(args.metatile_size)
+
+    bbox_west = -181
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_meta_tiles.py:498]: > since either you are setting `use_tiles_coords_generator` to True and must provide all the sides, you probably don't need these defaults

Yeah, but this is just to avoid IDE linter warnings such as "bbox_south might be None"

File: batch-setup/make_rawr_tiles.py [hunk 10]
```diff
@@ -10,6 +10,16 @@
...
+import sys
+import os
+from distutils.util import strtobool
+
+currentdir = os.path.dirname(os.path.realpath(__file__))
+parentdir = os.path.dirname(currentdir)
+sys.path.append(parentdir)
+
+from utils.tiles import BoundingBoxTilesCoordinateGenerator
+from utils.tiles import TilesCoordinateGenerator
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_rawr_tiles.py:22]: > are you using this import?

yup in the main function

File: batch-setup/make_rawr_tiles.py [hunk 11]
```diff
@@ -31,29 +41,31 @@ def all_tiles_at(zoom):
...
-                    present.add(coord)
-
-    missing = set(all_tiles_at(zoom)) - set(present)
-    return missing
+    if bool(tiles_coords_generator):
+        return set([c for c in tiles_coords_generator.
+                   generate_tiles_coordinates([zoom])])
+    else:
+        present = set()
+        finder = MissingTileFinder(
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_rawr_tiles.py:51]: the type of change is for backward compatible, but I make it an argument without default value since our code base is not large and I change all the places.

File: batch-setup/provision.sh [hunk 12]
```diff
@@ -138,6 +138,28 @@ python -u /usr/local/src/tileops/batch-setup/make_meta_tiles.py --date-prefix \$
...
+# stop on error
+set -e
+# echo commands before executing them (useful to check that the arguments are correct)
+set -x
+
+python -u /usr/local/src/tileops/batch-setup/make_tiles.py --num-db-replicas $NUM_DB_REPLICAS --max-vcpus $MAX_VCPUS $RUN_ID --missing-bucket $MISSING_BUCKET --meta-date-prefix $META_DATE_PREFIX $RAWR_BUCKET $META_BUCKET $DB_PASSWORD --overrides $JOB_ENV_OVERRIDES
+
+python -u /usr/local/src/tileops/batch-setup/make_rawr_tiles.py --config enqueue-rawr-batch.config.yaml --key-format-type hash-prefix --use-tiles-coords-generator --coords-generator-bbox-west $BBOX_WEST --coords-generator-bbox-south $BBOX_SOUTH --coords-generator-bbox-east $BBOX_EAST --coords-generator-bbox-north $BBOX_NORTH $RAWR_BUCKET $RUN_ID $MISSING_BUCKET
+
+python -u /usr/local/src/tileops/batch-setup/make_meta_tiles.py --date-prefix $META_DATE_PREFIX --missing-bucket $MISSING_BUCKET --key-format-type hash-prefix --metatile-size $METATILE_SIZE --use-tiles-coords-generator --coords-generator-bbox-west $BBOX_WEST --coords-generator-bbox-south $BBOX_SOUTH --coords-generator-bbox-east $BBOX_EAST --coords-generator-bbox-north $BBOX_NORTH $RAWR_BUCKET $META_BUCKET $RUN_ID
```
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/provision.sh:158]: > You'll need to update here if you update the bounding box spec

what do you mean?

File: utils/tiles.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    """
+    Generate tiles EPSG:3857 coordinates from a list of zooms within a range
+    """
+    def __init__(self, min_zoom, max_zoom):
+        # type: (int, int) -> None
+        """
+        :param min_zoom: the minimum zoom(inclusive) it can generate tiles for
+        :param max_zoom: the maximum zoom(inclusive) it can generate tiles for
+        """
+        self.min_zoom = min_zoom
```
[the remark this answers is not in the document]
Comment (peitili) [reply utils/tiles.py:20]: > maybe assert min_zoom <= max_zoom? Some things get weird below otherwise

fixed in the next commit.

File: batch-setup/make_meta_tiles.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (travisgrigsby) [reply batch-setup/make_meta_tiles.py:384]: awesome

File: batch-setup/make_meta_tiles.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (travisgrigsby) [reply batch-setup/make_meta_tiles.py:491]: Hmmm.  I don't love that we're including this code to prevent a thing we know can't happen.  I guess do whatever you want there, but this feels questionable

File: batch-setup/make_rawr_tiles.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (travisgrigsby) [reply batch-setup/make_rawr_tiles.py:22]: Oh I only saw the usage of the BoundingBox version

File: utils/tiles.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (travisgrigsby) [reply utils/tiles.py:10]: ok, just making sure

File: utils/tiles.py (hunk 3 above)
Comment (travisgrigsby) [reply utils/tiles.py:4]: you'll have to show me what you mean!  This sounds interesting

File: batch-setup/make_rawr_tiles.py [hunk 14]
```diff
@@ -219,6 +239,12 @@ def make_rawr_tiles(rawr_config_file, missing_config_file, missing_bucket,
                         "written out by make_tiles.py")
     parser.add_argument('missing_bucket', help="Bucket to store tile "
                         "enumeration logs in while calculating missing tiles.")
+    parser.add_argument('--use-tiles-coords-generator', type=lambda x: bool(strtobool(x)), nargs='?',
+                        const=True, default=False)
+    parser.add_argument('--coords-generator-bbox-west', type=float, help="west longitude of bounding box to build")
+    parser.add_argument('--coords-generator-bbox-south', type=float, help="south latitude of bounding box to build")
+    parser.add_argument('--coords-generator-bbox-east', type=float, help="east longitude of bounding box to build")
+    parser.add_argument('--coords-generator-bbox-north', type=float, help="north latitude of bounding box to build")
```
Comment (travisgrigsby) [inline batch-setup/make_rawr_tiles.py:247]: sorry I wrote this comment yesterday but didn't submit for some reason -  I think these should all be one arg.
`--coords-generator-bbox=1,2,3,4`
you can just split on commas and assert 4 floats or whatevs.  
https://wiki.openstreetmap.org/wiki/Bounding_Box

File: batch-setup/make_rawr_tiles.py [hunk 15]
```diff
@@ -232,6 +258,26 @@ def make_rawr_tiles(rawr_config_file, missing_config_file, missing_bucket,
               "AWS_DEFAULT_REGION to be set.")
         sys.exit(1)
 
+    bbox_west = -181
```
[the remark this answers is not in the document]
Comment (nvkelso) [reply batch-setup/make_rawr_tiles.py:261]: Please specify a valid bbox (within ±180, ±90), that is relatively small / not whole world. Like over Seattle / Western Washington state.

File: batch-setup/make_meta_tiles.py [hunk 16]
```diff
@@ -433,6 +458,12 @@ def render(self, num_retries, lense):
     parser.add_argument('--allowed-missing-tiles', default=2, type=int,
                         help='The maximum number of missing metatiles allowed '
                         'to continue the build process.')
+    parser.add_argument('--use-tiles-coords-generator', type=lambda x: bool(strtobool(x)), nargs='?',
```
Comment (iandees) [inline batch-setup/make_meta_tiles.py:461]: You can use action=store_true here instead of this. See argparse docs (https://docs.python.org/3/library/argparse.html#action) for more details.

File: batch-setup/make_meta_tiles.py [hunk 17]
```diff
@@ -433,6 +458,12 @@ def render(self, num_retries, lense):
...
+                        const=True, default=False)
+    parser.add_argument('--coords-generator-bbox-west', type=float, help="west longitude of bounding box to build")
```
Comment (iandees) [inline batch-setup/make_meta_tiles.py:463]: Agree that this should be a single string arg and then a split plus assert that there are four comma-separated floats. Also if you split and assert you don't have to worry about defaults.

File: batch-setup/make_rawr_tiles.py (hunk 14 above)
Comment (nvkelso) [reply batch-setup/make_rawr_tiles.py:247]: See what we do in Tilequeue here:

- https://github.com/tilezen/tilequeue/blob/33a4dd42e8f764bcc9817e89554ee21922f501a8/tilequeue/config.py#L62-L72

File: batch-setup/make_rawr_tiles.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_rawr_tiles.py:261]: > Please specify a valid bbox (within ±180, ±90), that is relatively small / not whole world. Like over Seattle / Western Washington state.

I will remove this section, I think people need to pass this information to trigger the bbox rebuild, rather than we rebuild a default bbox?

File: batch-setup/make_rawr_tiles.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_rawr_tiles.py:22]: > Oh I only saw the usage of the BoundingBox version

I don't think I understand this comment? are you saying we shouldn't import it globally? Importing globally vs inside function is for speed optimization, however in our case, I don't think it makes significant difference on the importing, by importing at the top it just improves the readability, later I will add a linter so that all the imports will be sorted.

File: utils/tiles.py (hunk 3 above)
Comment (peitili) [reply utils/tiles.py:4]: I already removed the lines in code. 

But what I meant is this, see the following screenshot. The function signature and return type is highlighted on the IDE in green background because it recognized the typing information. However List and Iterator are yellow because it cannot recognize it because I removed the `from typing import Iterator`. The typing library is only available in python3, so even if I leverage typing highlighting during development, I have to comment them out since our runtime is on python2 and will throw an error.

In short, these import statements will be used for typing in python3.

Snip20210519_4

File: batch-setup/make_meta_tiles.py (hunk 16 above)
Comment (peitili) [reply batch-setup/make_meta_tiles.py:461]: > You can use action=store_true here instead of this. See argparse docs (https://docs.python.org/3/library/argparse.html#action) for more details.

yup, looks like the current implementation will support multiple ways to set args.use_tiles_coords_generate to `True`
such as

--use-tiles-coords-generator

--use-tiles-coords-generator True

--use-tiles-coords-generator true

--use-tiles-coords-generator 1

--use-tiles-coords-generator yes

File: batch-setup/make_meta_tiles.py (hunk 17 above)
Comment (peitili) [reply batch-setup/make_meta_tiles.py:463]: changed in the newest commit

File: batch-setup/make_rawr_tiles.py (hunk 14 above)
Comment (peitili) [reply batch-setup/make_rawr_tiles.py:247]: changed in the newest commit

File: batch-setup/make_meta_tiles.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_meta_tiles.py:491]: removed in the newest commit

File: batch-setup/make_rawr_tiles.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (peitili) [reply batch-setup/make_rawr_tiles.py]: or maybe you mean TilesCoordinateGenerator? 
that is used in the typing hint

**Review by travisgrigsby — APPROVED** [review]
looks great!  Excited to try this out.
